vm/src/eval.rs

```diff
@@ -277,17 +286,65 @@ pub fn eval_inst(vm: &mut NexusVM<impl Memory>) -> Result<()> {
         ECALL { rd } => {
             RD = rd;
             vm.Z = vm.syscalls.syscall(vm.regs.pc, vm.regs.x, &vm.mem)?;
+
+            // Collects label from syscall and starts counting cycles
+            if vm.Z == (SyscallCode::ProfileCycles as u32) {
+                if let Some(s) = vm.syscalls.get_label() {
+                    let label = std::str::from_utf8(&s).unwrap();
+                    let fn_name = label.split(':').last().unwrap().to_string();
+
+                    match label.chars().next() {
+                        Some('^') => {
+                            // Start of the benchmark label
+                            vm.cycle_tracker
+                                .entry(fn_name)
+                                .or_insert((vm.global_clk, 0))
+                                .1 += 1;
+                        }
+                        Some('$') => {
+                            // End of the benchmark label
+                            // Write the total clock cycles this function takes.
+                            let (clk, counter) = vm
+                                .cycle_tracker
+                                .get_mut(&fn_name)
+                                .ok_or(NexusVMError::BenchmarkLabelError)?;
+                            *counter -= 1;
+                            if *counter == 0 {
+                                *clk = vm.global_clk - *clk;
+                            }
+                        }
+                        _ => (),
+                    }
+                }
+            }
         }
         UNIMP => {
             PC = vm.inst.pc;
         }
     }
 
+    // Counts cycles per instruction kind.
+    // In RISC-V:
+    // - Memory instructions are 3 cycles
+    // - Branch instructions are 2 cycles
+    // - System call instruction are 4 cycles and higher
+    // - ALU/ALUI/JAR/LUI/AUIPC instructions are 1 cycle
+    // - Unknown instruction is 0 cycle
+    let cycles = match vm.inst.inst {
+        LOAD { .. } | STORE { .. } => 3,
+        BR { .. } => 2,
+        LUI { .. } | AUIPC { .. } | JAL { .. } | JALR { .. } | ALUI { .. } | ALU { .. } => 1,
+        // Let's say ECALL is 4 cycles
```

Comment:
I suggest it should be rigorous instead of "Let's say". Optionally, we can skip counting cycles for `profile` syscall to eliminate the issue with build types.

Reply:
Removed in [6dd69d0](https://github.com/nexus-xyz/nexus-zkvm/pull/243/commits/6dd69d0f13e127f638cd2f061f65cf167686c5e9)